rack-protection/lib/rack/protection/ip_spoofing.rb

```diff
@@ -14,14 +14,27 @@ class IPSpoofing < Base
       default_reaction :deny
 
       def accepts?(env)
-        return true unless env.include? 'HTTP_X_FORWARDED_FOR'
+        return true unless env.include?('HTTP_X_FORWARDED_FOR') || env.include?('HTTP_FORWARDED')
+
+        ips = if env['HTTP_FORWARDED']
+          parse_forwarded(env['HTTP_FORWARDED'])
+        else
+          env['HTTP_X_FORWARDED_FOR'].split(',').map(&:strip)
+        end
 
-        ips = env['HTTP_X_FORWARDED_FOR'].split(',').map(&:strip)
         return false if env.include?('HTTP_CLIENT_IP') && (!ips.include? env['HTTP_CLIENT_IP'])
         return false if env.include?('HTTP_X_REAL_IP') && (!ips.include? env['HTTP_X_REAL_IP'])
 
         true
       end
+
+      private
+
+      def parse_forwarded(forwarded_header)
+        return [] unless forwarded_header_hash = Rack::Utils.forwarded_values(forwarded_header)
+
+        (forwarded_header_hash.fetch(:for, []) + forwarded_header_hash.fetch(:by, [])).map { |ip| ip.gsub(/\[|\]/, "")  }
+      end
```

Comment:
I think we should use the `#forwarded_for` method from `Rack::Request`, it will handled both `X-Forwarded-For` (when `Forwarded` isn't present)  and `Forwarded for=...`). It is unfortunate that there's no such helper for `by` though.

I don't think the `gsub` here is working correctly:

```ruby
irb(main):001> "[2001:db8:cafe::17]:4711".gsub(/\[|\]/, "")
=> "2001:db8:cafe::17:4711"
```

---

Just wanted to documented how I tested

I started a simple Rack application like this: `echo 'lowlevel_error_handler { |err| puts err.full_message }; require "rack"; app { |env| req=::Rack::Request.new(env); forwarded_values=::Rack::Utils.forwarded_values(env["HTTP_FORWARDED"]); [200, {}, ["#{req.forwarded_for}\n#{forwarded_values}"]] }' | puma --config /dev/stdin --port 4567 --log-requests`

And tried various requests

```shell
$ curl localhost:4567 -H 'Forwarded: By="[2001:db8:cafe::17]:4711"'

{:by=>["[2001:db8:cafe::17]:4711"]}
```
```
$ curl localhost:4567 -H 'Forwarded: by=192.0.2.60;proto=http;by=203.0.113.43'

{:by=>["192.0.2.60", "203.0.113.43"], :proto=>["http"]}
```
```
$ curl localhost:4567 -H 'Forwarded: For="[2001:db8:cafe::17]:4711"'
["2001:db8:cafe::17"]
{:for=>["[2001:db8:cafe::17]:4711"]}
```